File: Convert Sorted List to Binary Search Tree.cpp
1차 22-04-14
Time: 19분 20초
Success
Details 
Runtime: 16 ms, faster than 99.52% of C++ online submissions for Convert Sorted List to Binary Search Tree.
Memory Usage: 28.4 MB, less than 32.85% of C++ online submissions for Convert Sorted List to Binary Search Tree.
    
2차 22-06-09
코틀린으로 풀이하였음. 
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    /*
    문제:
    "링크드리스트"가 주어진다. 오름차순으로 정렬되어있다.
    balanced binary search tree로 만들어라.
    
    제한:
    
    해결방법:
    각 재귀마다 왼쪽 오른쪽 나누어서 돌린다.
    단, size라는 변수를 두어 mid를 찾기 쉽게 한다.
    또한 base case는 size로 비교하면 size가 네개일 때 재귀를 돌리다보면 size는 1인데 leftTree의 노드가 두개인 경우가 발생한다.(ex -10, 4, 5, 6)
    그래서 l과 r로 바로 비교한다. 그게 더 정확하다.
    */
    TreeNode* makeTree(ListNode* l, ListNode* r, int size){
        if(l == r)
            return new TreeNode(l->val);
        if(l->next == r){
            TreeNode *ret = new TreeNode(l->val);
            TreeNode* right = new TreeNode(r->val);
            ret->right = right;
            return ret;
        }
        ListNode* mid =l;
        
        ListNode *beforeMid, *afterMid;
        for(int i=0; i<size/2-1; i++)
            mid = mid->next;
        
        
        beforeMid = mid;
        mid = mid->next;
        TreeNode* midTreeNode = new TreeNode(mid->val);
        afterMid = mid->next;
        midTreeNode->left = makeTree(l, beforeMid, (size-1)/2);
        midTreeNode->right = makeTree(afterMid, r, (size-1)/2);
        
        return midTreeNode;
    }
    TreeNode* sortedListToBST(ListNode* head) {
        if(head == NULL)
            return NULL;
        int listSize = 0;
        ListNode* l = head;
        ListNode* r;
        while(head->next!=NULL)
            listSize++, head=head->next;
        r = head;
        cout << l->val << ' ' << r->val << ' ' << listSize+1 << endl;
        return makeTree(l, r, listSize+1);
    }
};

File: kotlin/Convert Sorted List to Binary Search Tree.kt
2차 22-06-09
회고: 아이디어 떠올리는데 성공. 자세한 구현도 해봄. 그러나 구현에서 실패함
아래 코드는 Discussion에서 퍼온 것.

🎈내 풀이
1)기존에 내가 구현했던 코드는 turtle & hale이라는, 리스트의 중간을 찾아내는 알고리즘으로 리스트의 중간을 찾아냈었음.
2)그래서 각각 리스트의 처음과 중간, 중간과 끝을 매개변수로 주어서 재귀로 할려고 했음

🎈알아야 할 점
NUll 체크 할 때, 어떤건 !!을 쓰고, 어떤건 ?을 쓰고, 어떤건 ?:을 쓴다??
이슈에 첨부하겠음.

fun sortedListToBST(head: ListNode?): TreeNode? {
        val N = size(head) // size of LL
        if (N == 0) return null
        if (N == 1) return TreeNode(head!!.`val`)
        val A = IntArray(N) // LL data in array
        populate(A, head) // fill the array
        return construct(head, A, 0, N - 1)
    }
    
    private fun construct(head: ListNode?, A: IntArray, start: Int, end: Int): TreeNode? {
        if (start > end) return null
        val mid = (start + end) / 2
        val root = TreeNode(A[mid])
        root.left = construct(head, A, start, mid - 1)
        root.right = construct(head, A, mid + 1, end)
        return root
    }
    
    private fun size(head: ListNode?): Int {
        if (head == null) return 0
        if (head.next == null) return 1
        var ans = 0
        var node = head
        while (node != null) {
            ans++
            node = node.next
        }
        return ans
    }
    
    private fun populate(A: IntArray, head: ListNode?) {
        var node = head
        for (i in A.indices) {
            A[i] = node!!.`val`
            node = node?.next
        }
    }
